diffusion/models/autoencoder.py

```diff
@@ -225,6 +226,52 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
         return h
 
 
+class GaussianDistribution:
+    """Gaussian distribution parameterized with mean and log variance."""
+
+    def __init__(self, mean: torch.Tensor, log_var: torch.Tensor):
+        self.mean = mean
+        self.log_var = log_var
+        self.var = torch.exp(log_var)
+        self.std = torch.exp(0.5 * log_var)
+
+    def __getitem__(self, key):
+        if key == 'latent_dist':
+            return GaussianDistribution(self.mean, self.log_var)
+        elif key == 'mean':
+            return self.mean[key]
+        elif key == 'log_var':
+            return self.log_var[key]
+        else:
+            raise KeyError(key)
+
+    @property
+    def latent_dist(self):
+        return self
+
+    def sample(self) -> torch.Tensor:
+        """Sample from the distribution."""
+        return self.mean + self.std * torch.randn_like(self.mean)
+
+
+class AutoEncoderOutput:
+    """Output from an autoencoder."""
+
+    def __init__(self, x_recon: torch.Tensor):
+        self.x_recon = x_recon
+
+    def __getitem__(self, key):
+        if key == 'x_recon':
+            return self.x_recon
+        else:
+            raise KeyError(key)
+
+    @property
+    def sample(self) -> torch.Tensor:
+        """Sample from the output."""
+        return self.x_recon
```

Comment:
It is used in our generate function to decode from generated latents [here](https://github.com/mosaicml/diffusion/blob/90b97ed93e42cde98bc536501fd9ef16bb9a9ced/diffusion/models/stable_diffusion.py#L541C17-L541C17). I agree it is a bit weird, and we could modify our `generate` to not use `sample` like this. But this is to make things a bit more compatible with the HF autoencoder. There _could_ be some randomness here in sampling from the VAE decoder, but it would degrade performance.

Reply:
Thanks!